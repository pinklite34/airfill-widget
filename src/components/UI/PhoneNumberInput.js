import React, {Component} from 'react';
import styled from 'styled-components';
import { parse, format } from 'libphonenumber-js';
import Field from './Field';

const PhoneNumberField = styled(Field)`
  .refill-field {
    display: flex;
    justify-content: flex-start;
  }
  .refill-number-field-input {
    flex: 1 1 auto;
    width: auto;
    min-width: 100px;
    max-width: 20em;
    margin-right: 8px;
  }
`

const getErrorMessage = error => {
  const messages = {
    'Country not supported':
      'We\'re sorry, but this country is not supported right now.',
    'Country code needed':
      'Please enter the number including a country code, starting with \'+\'.',
    'Phone number entered is too short':
      'The number you entered is too short. Are you sure it is correct?'
  };
  if (error in messages) {
    return messages[error];
  } else if (error && error.indexOf('Number is not valid') !== -1 || error === true) {
    return 'This is not a valid phone number. Make sure it is correct and try again!';
  }
  return error;
};


class PhoneNumberInput extends Component {
  getCountryCode = () => {
    const {country} = this.props;
    return country ? country.countryCallingCodes[0] : undefined;
  }

  getAlpha2 = () => {
    const { country={} } = this.props;
    const { alpha2=null } = country;
    return alpha2;
  }

  formatDefaultValue = () => {
    let { defaultValue = '' } = this.props;

    const cc = this.getCountryCode();
    const alpha2 = this.getAlpha2();
    const number = defaultValue.replace(/[^\d\+]/, '');
    const validNumber = this.validateNumber(number);

    defaultValue = validNumber
      ? format(validNumber, alpha2, 'International')
      : number;

    return defaultValue || `${cc} `;
  }

  validateNumber = (number) => {
    const cc = this.getCountryCode();
    const alpha2 = this.getAlpha2();

    if (!number) {
      return null;
    }

    // Try to validate number
    const { country, phone } = parse(number, alpha2);

    if (phone && country) {
      // Number is valid, return it
      return phone;
    } else if (number.charAt(0) !== '+') {
      // Number lacks country code - add it and try again
      return this.validateNumber(cc + number);
    }

    // Number is not valid
    return null
  }

  handleChange = (event) => {
    const number = event.target.value.replace(/[^\d\+]/, '');

    const validNumber = this.validateNumber(number);
    const {country: { alpha2 }} = this.props;

    if (this.props.onChange) {
      this.props.onChange(validNumber ?
        format(validNumber, alpha2, 'International') : number
      )
    }
  }


  render() {
    const {label, hint, error, children} = this.props;

    return (
      <PhoneNumberField
        className="refill-number-field"
        label={label || 'Phone number'}
        hint={hint || 'The phone number to top up'}
        error={getErrorMessage(error)}
      >
        <input
          className="refill-number-field-input"
          type="tel"
          onChange={this.handleChange}
          defaultValue={this.formatDefaultValue()}
        />
        {children}
      </PhoneNumberField>
    );
  }
}

export default PhoneNumberInput;
