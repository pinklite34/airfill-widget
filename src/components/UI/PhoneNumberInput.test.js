import React from 'react';
import renderer from 'react-test-renderer';
import PhoneNumberInput from './PhoneNumberInput';

const props = {
  onCahnge: jest.fn(),
  type: null,
  country: { alpha2: 'US', countryCallingCodes: ['+1']},
  defaultValue: '',
  label: null,
  hint: null,
  error: null,
  children: '--CHILDREN--'
};

it('renders correctly', () => {
  const tree = renderer.create(<PhoneNumberInput {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Treats VOIP as phone number', () => {
  const tree = renderer.create(<PhoneNumberInput {...props} type="VOIP" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Treats DHT as account number', () => {
  const tree = renderer.create(<PhoneNumberInput {...props} type="DHT"/>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Uses default value', () => {
  const tree = renderer.create(<PhoneNumberInput {...props} defaultValue="+12015555555" />).toJSON();
  expect(tree).toMatchSnapshot();
});

