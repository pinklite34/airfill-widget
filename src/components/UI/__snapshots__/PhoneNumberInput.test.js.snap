// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treats DHT as account number 1`] = `
<div
  className="refill-number-field fYeEn hulnrW"
>
  <label
    className="refill-field-label"
    htmlFor={undefined}
  >
    Account number
  </label>
  <div
    className="refill-field"
  >
    <input
      className="refill-number-field-input"
      defaultValue=""
      onChange={[Function]}
      type="tel"
    />
    --CHILDREN--
  </div>
  <p
    className="refill-field-hint"
    dangerouslySetInnerHTML={
      Object {
        "__html": "The DHT account number to top up",
      }
    }
  />
</div>
`;

exports[`Treats VOIP as phone number 1`] = `
<div
  className="refill-number-field fYeEn hulnrW"
>
  <label
    className="refill-field-label"
    htmlFor={undefined}
  >
    Phone number
  </label>
  <div
    className="refill-field"
  >
    <input
      className="refill-number-field-input"
      defaultValue="+1 "
      onChange={[Function]}
      type="tel"
    />
    --CHILDREN--
  </div>
  <p
    className="refill-field-hint"
    dangerouslySetInnerHTML={
      Object {
        "__html": "The phone number to top up",
      }
    }
  />
</div>
`;

exports[`Uses default value 1`] = `
<div
  className="refill-number-field fYeEn hulnrW"
>
  <label
    className="refill-field-label"
    htmlFor={undefined}
  >
    Phone number
  </label>
  <div
    className="refill-field"
  >
    <input
      className="refill-number-field-input"
      defaultValue="+1 201 555 5555"
      onChange={[Function]}
      type="tel"
    />
    --CHILDREN--
  </div>
  <p
    className="refill-field-hint"
    dangerouslySetInnerHTML={
      Object {
        "__html": "The phone number to top up",
      }
    }
  />
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="refill-number-field fYeEn hulnrW"
>
  <label
    className="refill-field-label"
    htmlFor={undefined}
  >
    Phone number
  </label>
  <div
    className="refill-field"
  >
    <input
      className="refill-number-field-input"
      defaultValue="+1 "
      onChange={[Function]}
      type="tel"
    />
    --CHILDREN--
  </div>
  <p
    className="refill-field-hint"
    dangerouslySetInnerHTML={
      Object {
        "__html": "The phone number to top up",
      }
    }
  />
</div>
`;
